### PR TITLE
Make unimplemented paths defined in schema 501

### PR DESCRIPTION
**Pull request by dcolligan**
As of schema version 84927c4cc9a74a3db310026de8b31cdca8bbcfd0

Issue #431 

It would probably be better to use the flask route syntax instead of the `re` module, but I'm not sure if Flask supports matching routes outside an app context...

### CODE REVIEW

**Comment by jeromekelleher** [comment]
This looks good, but I'm a bit worried about the regex stuff. I know it's a bit tedious, but would it be simpler to just put in the handler functions and 501 from all the ones that we haven't implemented?

**Comment by dcolligan** [comment]
@jeromekelleher alright, did that (which required a little flask hacking)

File: tests/unit/test_views.py [hunk 1]
```diff
@@ -269,3 +272,46 @@ def testCurrentVersion(self):
...
+            '/mode/<id>',
+            '/datasets/<id>',
+            '/readgroupsets/<id>',
+            '/readgroups/<id>',
+        ]
+
+        def runRequest(method, path):
+            requestPath = path.replace('<id>', 'someId')
+            versionedPath = utils.applyVersion(requestPath)
+            print(method, versionedPath)
```
Comment (jeromekelleher) [inline tests/unit/test_views.py]: Stray print here.

File: ga4gh/frontend.py [hunk 2]
```diff
@@ -384,8 +403,139 @@ def searchDatasets(version):
         version, flask.request, app.backend.searchDatasets)
 
 
+# The below paths have not yet been implemented
+
+
+@app.route('/<version>/callsets/<no(search):id>')
```
Comment (jeromekelleher) [inline ga4gh/frontend.py]: I don't get it: why the <no(search):id> bit?

Comment (dcolligan) [inline ga4gh/frontend.py]: If there are two routes, 
```
/<version>/callsets/search
/<version>/callsets/<id>
```
a request to `/someVersion/callsets/search` will be routed to the later, which is not what we want.  Therefore, we have this converter to explicitly say that the `<id>` attribute of the later should not match on a literal `search`, thus routing the request correctly.

**Comment by dcolligan** [comment]
@jeromekelleher pushed update removing print statement

File: ga4gh/frontend.py (hunk 2 above)
Comment (jeromekelleher) [inline ga4gh/frontend.py]: Ugh, this is pretty horrible. So, we can never have an object with ID = 'search'? Is this a standard RESTish issue do you think, or something new we've made up ourselves?

There's not a lot we can do about it though. Can you add a quick comment on the Converter class to explain why we need this? I'm happy to merge then.

Comment (dcolligan) [inline ga4gh/frontend.py]: I don't think it's terribly unusual to have certain reserved paths in REST (in any event, it's defined by the schema, so our hands are tied somewhat).  Alternatives would be (if `foo` is the resource):

- have an attribute in the POST payload to `/foo` specifying the action (this seems like overloading paths = bad)
- have `/foo` for CRUD methods and `/foos` for more advanced ones over all the `foo` objects like search (but foo vs foos distinction could be seen as confusing)... that is, if we even ever implement CRUD methods
- organize the URL space differently and have a `/search/foos` endpoint (seems to break the object-then-action URL simplicity)
- have some other piece of data be sent along with the usual payload to `/foo`, like a header or something (confusing)
- just assume a POST to `/foo` means we want to search (possible, but would make any CRUD implementation backwards-incompatible)

None of these seem as attractive as what we're doing now... in any event, `search` seems like a pretty improbable ID (although, I don't doubt that someone, somewhere...)

**Comment by dcolligan** [comment]
@jeromekelleher pushed update with comment

**Comment by jeromekelleher** [comment]
Thanks @dcolligan, merging.

File: ga4gh/frontend.py (hunk 2 above)
Comment (jeromekelleher) [inline ga4gh/frontend.py]: Fair enough; thanks for the analysis.

**Pull request merged**